test/k8sT/DatapathConfiguration.go

```diff
@@ -645,9 +645,86 @@ var _ = Describe("K8sDatapathConfig", func() {
 			}, DeployCiliumOptionsAndDNS)
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
 		})
+
+		SkipItIf(helpers.RunsOnGKE, "With VXLAN", func() {
+			deploymentManager.DeployCilium(map[string]string{
+				"hostFirewall": "true",
+			}, DeployCiliumOptionsAndDNS)
+			testHostFirewall(kubectl)
+		})
+
+		// We need to skip this test when using kube-proxy because of #14859.
+		SkipItIf(helpers.RunsWithKubeProxy, "With native routing", func() {
+			options := map[string]string{
+				"hostFirewall":         "true",
+				"tunnel":               "disabled",
+				"autoDirectNodeRoutes": "true",
+			}
+			// We can't rely on gke.enabled because it enables
+			// per-endpoint routes which are incompatible with
+			// the host firewall.
+			if helpers.RunsOnGKE() {
+				options["gke.enabled"] = "false"
+			}
+			deploymentManager.DeployCilium(options, DeployCiliumOptionsAndDNS)
+			testHostFirewall(kubectl)
+		})
 	})
 })
 
+func testHostFirewall(kubectl *helpers.Kubectl) {
+	randomNs := deploymentManager.DeployRandomNamespaceShared(DemoHostFirewall)
+	deploymentManager.WaitUntilReady()
+
+	demoHostPolicies := helpers.ManifestGet(kubectl.BasePath(), "host-policies.yaml")
+	By(fmt.Sprintf("Applying policies %s", demoHostPolicies))
+	_, err := kubectl.CiliumPolicyAction(randomNs, demoHostPolicies, helpers.KubectlApply, helpers.HelperTimeout)
+	ExpectWithOffset(1, err).Should(BeNil(), fmt.Sprintf("Error creating resource %s: %s", demoHostPolicies, err))
+
+	defer func() {
+		By(fmt.Sprintf("Removing policies %s", demoHostPolicies))
+		_, err = kubectl.CiliumPolicyAction(randomNs, demoHostPolicies, helpers.KubectlDelete, helpers.HelperTimeout)
+		ExpectWithOffset(1, err).Should(BeNil(), fmt.Sprintf("Error removing resource %s: %s", demoHostPolicies, err))
+	}()
```

Comment:
I was somehow under the impression that we must avoid `defer` in the CI test code, because of the way that test failure assertions break the program flow (ie I'm not sure that this code will be guaranteed to run on failure as it throws an exception and the outer CI code catches it via `recover()`). Is there a reason you avoided adding this step as an `AfterEach`?

Reply:
I see `defer`s being used at many places in our CI tests. `defer` functions in Go are evaluated at the time of function returns, or panics. I'm not sure how CI test assertions could alter that.

Reply:
I've switched it anyway to use `AfterEach`. That allows us to delete CCNPs after every single host firewall test instead of doing it manually in each test. I think it's best because a leftover host policy can be pretty bad for the CI (failing all subsequent tests and potentially brick the nodes).